### PR TITLE
Change character controller to update on physics timestep rather than on tick

### PR DESCRIPTION
**Pull request by greerdv**
Signed-off-by: greerdv <greerdv@amazon.com>

## What does this PR do?
Changes character controller updates to happen per physics timestep, rather than per tick (should improve smoothness when riding on moving platforms).
Makes character controller IsOnGround check more consistent.

## How was this PR tested?
Made a level with moving platforms. Tested with script debug printing IsOnGround result.

### CODE REVIEW

**Comment by greerdv** [comment]
Old behaviour
https://user-images.githubusercontent.com/82226198/180849342-6195503a-31a4-4957-8614-f5043e71fa6a.mp4

**Comment by greerdv** [comment]
New behaviour
https://user-images.githubusercontent.com/82226198/180849474-45f3f4c6-0bf4-4d60-980f-f855f9d4d02e.mp4

**Review by AMZN-puvvadar — APPROVED** [review]
The changes around the Multiplayer Gem seem reasonable to me.

File: Code/Framework/AzFramework/AzFramework/Physics/PhysicsScene.h [hunk 1]
```diff
@@ -485,12 +485,12 @@ namespace AzPhysics
 
     inline void Scene::RegisterSceneSimulationStartHandler(SceneEvents::OnSceneSimulationStartHandler& handler)
     {
-        handler.Connect(m_sceneSimuationStartEvent);
+        handler.Connect(m_sceneSimulationStartEvent);
```
Comment (hultonha) [inline Code/Framework/AzFramework/AzFramework/Physics/PhysicsScene.h:488]: Good spot! 😅

File: Code/Framework/AzFramework/AzFramework/Physics/Character.h [hunk 2]
```diff
@@ -112,24 +112,60 @@ namespace Physics
...
+        /// can be combined from separate calls to AddVelocityForTick or AddVelocityForPhysicsTimestep.
+        /// Accumulating the requests avoids performing multiple expensive character updates, and avoids any effects from the order of
+        /// requests within a tick. Users who wish to add a new movement effect should generally just be able to use AddVelocityForTick or
+        /// AddVelocityForPhysicsTimestep, and rely on the existing physics system call to ApplyRequestedVelocity.
         /// @param deltaTime The duration over which to apply the accumulated requested velocity.
         virtual void ApplyRequestedVelocity(float deltaTime) = 0;
 
+        /// Sets the accumulated velocity requests which last for the duration of a tick to zero.
+        /// The expected usage is for this funciton to be called internally by the physics system, to flush per tick
+        /// velocity requests once per tick. Velocity requests with duration of the physics timestep are not affected.
```
[the remark this answers is not in the document]
Comment (hultonha) [reply Code/Framework/AzFramework/AzFramework/Physics/Character.h:154]: question: Should any of these comments also mention the `maximumSpeed` value and the fact things will be clamped if exceeding that? Likely useful to know

File: Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterGameplayComponent.cpp [hunk 3]
```diff
@@ -35,6 +36,10 @@ namespace PhysX
                     ->DataElement(AZ::Edit::UIHandlers::Default, &CharacterGameplayConfiguration::m_gravityMultiplier,
                         "Gravity Multiplier", "Multiplier for global gravity value that applies only to this character entity.")
                     ->Attribute(AZ::Edit::Attributes::Step, 0.1f)
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &CharacterGameplayConfiguration::m_groundDetectionBoxHeight,
+                        "Ground Detection Box Height", "Vertical size of box to use when testing for ground contact.")
```
Comment (hultonha) [inline Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterGameplayComponent.cpp:40]: question: Where does this box start from? Would it be possible to expand the comment slightly to give a little more context?

File: Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterGameplayComponent.cpp [hunk 4]
```diff
@@ -93,25 +108,84 @@ namespace PhysX
...
+        if (auto* scene = character->GetScene())
+        {
+            // use a box shape for the overlap, even if the character geometry is a capsule, to avoid difficulties with
+            // the curved base of the capsule
+            AZ::Vector3 footBoxDimensions(0.0f, 0.0f, m_groundDetectionBoxHeight);
+            if (pxController->getType() == physx::PxControllerShapeType::eCAPSULE)
+            {
+                const float radius = static_cast<physx::PxCapsuleController*>(pxController)->getRadius();
+                footBoxDimensions.SetX(2.0f * radius);
+                footBoxDimensions.SetY(2.0f * radius);
```
Comment (hultonha) [inline Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterGameplayComponent.cpp:162]: suggestion: Could we combine the Set in one call (not sure what the type is here by can we just do `Set(radius * 2.0f, radius * 2.0f);`

**Review by hultonha — APPROVED** [review]
Change looks good, I have a few minor comments. One other question is I don't see any tests for this. Is this something we could verify in a simplified physics test (to capture the bad behavior and show that this mitigates/fixes it?). I understand this might be difficult but even factoring out part of the logic to test could be useful. If it is a nightmare to test it might still be worth making a GHI to go back to or making @FuzzyCarterAWS aware as maybe this is something we'd need a python test for in future?

**Comment by hultonha** [comment]
Could you please update the description for this PR too please to say Character Controller for the uninitiated 🙂

**Comment by greerdv** [comment]
> Could you please update the description for this PR too please to say Character Controller for the uninitiated 🙂

Good idea, done.

File: Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterGameplayComponent.cpp (hunk 4 above)
Comment (greerdv) [reply Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterGameplayComponent.cpp:162]: It's a Vector3. Guess I could do ... = Vector3(2.0f * radius, 2.0f * radius, m_groundDetectionBoxHeight)

File: Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterGameplayComponent.cpp (hunk 3 above)
Comment (greerdv) [reply Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterGameplayComponent.cpp:40]: OK, I'll elaborate on the comment.

File: Code/Framework/AzFramework/AzFramework/Physics/Character.h [hunk 5]
```diff
@@ -112,24 +112,60 @@ namespace Physics
         virtual AzPhysics::CollisionGroup GetCollisionGroup() const = 0;
         virtual AZ::Crc32 GetColliderTag() const = 0;
 
-        /// Queues up a request to apply a velocity to the character.
-        /// All requests received during a tick are accumulated (so for example, the effects of animation and gravity
-        /// can be applied in two separate requests), and a movement with the accumulated velocity is performed once
-        /// per tick, prior to the physics update.
+        // O3DE_DEPRECATION_NOTICE()
```
[the remark this answers is not in the document]
Comment (greerdv) [reply Code/Framework/AzFramework/AzFramework/Physics/Character.h:115]: Just made https://github.com/o3de/o3de/issues/10883, will add in the deprecation comment.

File: Code/Framework/AzFramework/AzFramework/Physics/Character.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (greerdv) [reply Code/Framework/AzFramework/AzFramework/Physics/Character.h:154]: Added a reference to the maximum speed setting to the doxygen markup.

File: Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h [hunk 6]
```diff
@@ -70,12 +70,29 @@ namespace Physics
         /// Gets the observed velocity of the character, which may differ from the desired velocity if the character is obstructed.
         virtual AZ::Vector3 GetVelocity() const = 0;
 
-        /// Queues up a request to apply a velocity to the character.
-        /// All requests received during a tick are accumulated (so for example, the effects of animation and gravity
-        /// can be applied in two separate requests), and a movement with the accumulated velocity is performed once
-        /// per tick, prior to the physics update.
+        // O3DE_DEPRECATION_NOTICE()
```
[the remark this answers is not in the document]
Comment (greerdv) [reply Code/Framework/AzFramework/AzFramework/Physics/CharacterBus.h:73]: Done

File: Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterGameplayComponent.cpp (hunk 4 above)
Comment (hultonha) [reply Gems/PhysX/Code/Source/PhysXCharacters/Components/CharacterGameplayComponent.cpp]: I think that's a bit nice but totally up to you 👍

File: Code/Framework/AzFramework/AzFramework/Physics/Character.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (hultonha) [reply Code/Framework/AzFramework/AzFramework/Physics/Character.h]: Thank you!

File: Code/Framework/AzFramework/AzFramework/Physics/Character.h [hunk 7]
```diff
@@ -112,24 +112,60 @@ namespace Physics
...
+        };
+
+        /// Queues up a request to apply a velocity to the character, lasting for the duration of the tick.
+        /// All requests received are accumulated (so for example, the effects of animation and gravity
+        /// can be applied in two separate requests), and the accumulated velocity is used when the character updates.
+        /// Velocities added this way will apply until the end of the tick.
         /// Obstacles may prevent the actual movement from exactly matching the requested movement.
-        /// @param velocity The velocity to be added to the accumulated requests.
-        virtual void AddVelocity(const AZ::Vector3& velocity) = 0;
+        /// @param velocity The velocity to be added to the accumulated requests, lasting for the duration of the tick.
```
[the remark this answers is not in the document]
Comment (hultonha) [reply Code/Framework/AzFramework/AzFramework/Physics/Character.h:131]: I just meant as newbie developer, when would you use this vs the other, is one preferred etc? Not a huge deal I just think I'd be a bit confused by what do to do get the best effect

File: Code/Framework/AzFramework/AzFramework/Physics/Character.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (hultonha) [reply Code/Framework/AzFramework/AzFramework/Physics/Character.h:154]: Thanks!

**Review by hultonha — APPROVED** [review]

**Pull request merged**